### PR TITLE
Porting TexlTests

### PR DESCRIPTION
**Pull request by jas-valgotar**
-Ports Texl Tests from [PAClient ](https://msazure.visualstudio.com/OneAgile/_git/PowerApps-Client?path=/src/Cloud/unittests/DocumentServer.Core.LargeTests/TexlTest.cs&_a=contents&version=GBmaster)to PowerFX

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by MikeStall — APPROVED** [review]
:shipit:

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs [hunk 1]
```diff
@@ -43,31 +40,65 @@ public void TexlDateOverloads_Negative(string script)
...
+            Assert.False(result.IsSuccess);
         }
 
         [Theory]
         [InlineData("DateAdd([Date(2000,1,1)],1)", "*[Value:d]")]
         [InlineData("DateAdd([Date(2000,1,1)],[3])", "*[Value:d]")]
         [InlineData("DateAdd(Table({a:Date(2000,1,1)}),[3])", "*[a:d]")]
         [InlineData("DateAdd(Date(2000,1,1),[1])", "*[Result:d]")]
+        [InlineData("DateAdd(DateTimeValue(\"1 Jan 2015\"), 2)", "d")]
+        [InlineData("DateAdd(DateTimeValue(\"1 Jan 2015\"), 2, TimeUnit!Years)", "d")]
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:86]: >! [](http://example.com/codeflow?start=71&length=1)

Even though this is supported (`!` as the field selector), let's not have tests showing this. We want the tests to be used as a "template" for people to see how the language can be used.

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs [hunk 2]
```diff
@@ -43,31 +40,65 @@ public void TexlDateOverloads_Negative(string script)
...
+        {
+            Assert.True(DType.TryParse(expectedType, out var type));
+            Assert.True(type.IsValid);
+
+            TestSimpleBindingSuccess(script, type);
+        }
+
+        [Theory]
+        
+        [InlineData("DateAdd([Date(2000,1,1)],1)", "*[Value:d]")]
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:99]: [nit] remove empty line

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs (hunk 1 above)
Comment (MikeStall) [reply src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:86]: Although ... we're still not allowed to break it, so we still need at least one test for it.  This PR is capturing existing tests that we have in PA client and run closer to fx core.

Comment (CarlosFigueiraMSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:86]: We can have one or a few _specific_ tests that use this syntax for back-compat purposes, but it shouldn't be sprinkled throughout the codebase.

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs [hunk 3]
```diff
@@ -117,5 +148,2026 @@ public void TexlDateTableFunctions_Negative(string expression)
...
+                symbol);
+
+            TestSimpleBindingSuccess(
+                "Trunc(X, 4)",
+                TestUtils.DT("*[Nnnuuummm:n]"),
+                symbol);
+        }
+
+        [Fact]
+        public void TexlFunctionTypeSemanticsRound()
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:853]: >TexlFunctionTypeSemanticsRound [](http://example.com/codeflow?start=20&length=30)

We should have a separate test that validates that the output is *[Value:n] if the ConsistentOneColumnTableResult feature is enabled. Same for other tabular functions

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs [hunk 4]
```diff
@@ -117,5 +147,2211 @@ public void TexlDateTableFunctions_Negative(string expression)
...
+        }
+
+        [Fact]
+        public void TexlFunctionTypeSemanticsTruncOneParam_ConsistentOneColumnTableResult()
+        {
+            var symbol = new SymbolTable();
+            symbol.AddVariable("T", new TableType(TestUtils.DT("*[A:n]")));
+            TestSimpleBindingSuccess(
+                "Trunc(T)",
+                TestUtils.DT("*[A:n]"),
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:905]: >*[A:n] [](http://example.com/codeflow?start=30&length=6)

If we're using the ConsistentOneColumnTableResult flag, it should be `*[Value:n]`

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs [hunk 5]
```diff
@@ -117,5 +147,2211 @@ public void TexlDateTableFunctions_Negative(string expression)
...
+        [InlineData("Round(X, 4)", "*[Nnnuuummm:n]")]
+        public void TexlFunctionTypeSemanticsRound_ConsistentOneColumnTableResult(string expression, string expectedType)
+        {
+            var symbol = new SymbolTable();
+            symbol.AddVariable("T", new TableType(TestUtils.DT("*[digits:n]")));
+            symbol.AddVariable("X", new TableType(TestUtils.DT("*[Nnnuuummm:n]")));
+
+            TestSimpleBindingSuccess(
+                expression,
+                TestUtils.DT(expectedType),
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:976]: >expectedType [](http://example.com/codeflow?start=29&length=12)

It should always be `*[Value:n]` if we're using the ConsistentOneColumnTableResult flag

Comment (jas-valgotar) [reply src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:976]: So, is it okay to remove the third test, Round(X,4)? I just had it to show difference of having table on arg0 vs arg1

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs [hunk 6]
```diff
@@ -117,5 +147,2211 @@ public void TexlDateTableFunctions_Negative(string expression)
...
+                script,
+                TestUtils.DT(expectedType),
+                symbol);
+        }
+
+        [Theory]
+        [InlineData("Sqrt(1234.567)", "n")]
+        [InlineData("Sqrt(T)", "*[A:n]")]
+        [InlineData("Sqrt(T2!Value)", "*[Value:n]")]
+        public void TexlFunctionTypeSemanticsSqrt(string script, string expectedType)
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:1072]: >TexlFunctionTypeSemanticsSqrt [](http://example.com/codeflow?start=20&length=29)

There are some other functions (like Sqrt) for which we should have tests for the ConsistentOneColumnTableResult flag. Basically every function on libraries\Microsoft.PowerFx.Core\Texl\BuiltinFunctionsCore.cs with a "T" suffix should have tests for that.

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs (hunk 4 above)
Comment (jas-valgotar) [reply src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:905]: The current code returns the same return type as arg0, is that not expected behavior?

Comment (CarlosFigueiraMSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:905]: No, the current code is wrong (we found a bug, great!) The "ConsistentOneColumnTableResult" flag will change the schema of all functions that return a single-column table (tabular overloads of scalar functions) to always have a *[Value:<type>] schema.

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs (hunk 5 above)
Comment (CarlosFigueiraMSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:976]: No, it's good to have the third test - we want to make sure that, with that flag on, we have the same schema for the result in all cases.

Comment (jas-valgotar) [reply src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:976]: Then we have a Bug here too! Coz, Round function returns return type same as arg0 if arg0 is table.

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs (hunk 1 above)
Comment (jas-valgotar) [reply src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:86]: removed tests with field selector, while keeping just few.

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs (hunk 5 above)
Comment (jas-valgotar) [reply src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:928]: I have filed issue https://github.com/microsoft/Power-Fx/issues/709 for this bug.

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs (hunk 4 above)
Comment (jas-valgotar) [reply src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs]: I have filed issue https://github.com/microsoft/Power-Fx/issues/709 for this bug.

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs (hunk 6 above)
Comment (jas-valgotar) [reply src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs:1021]: This will be tracked via https://github.com/microsoft/Power-Fx/issues/710

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by MikeStall — APPROVED** [review]
:shipit:

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by CarlosFigueiraMSFT — APPROVED** [review]
:shipit:

**Pull request merged**